power: add DEVICE_PM_LOW_POWER_STATE for device power management

wentongwu opened the pull request:
when system going to sleep state(not deep sleep), make peripherals
go to DEVICE_PM_LOW_POWER_STATE state which need less time than
DEVICE_PM_SUSPEND_STATE to save more power in the sleep state.

Fixes: #17838.

wentongwu committed:
Style: remove extern on the function declarations in the header

Remove extern on the function declarations in the header file to
keep consistent with others.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
@pabigot @vanti @mnkp I think I need consider more for the centralized and distributed device power management, so close #18007. And based on current policy, I want to add DEVICE_PM_LOW_POWER_STATE for device power management to save more power, please give a review for this PR, thanks a lot.

pabigot reviewed (approved):
This seems like the right move: "sleep state" is distinguished from "deep sleep state" by preserving peripheral configuration, so if system isn't going to deep sleep it seems to be OK to move device to its "low-power context-preserved" state.

vanti reviewed (changes requested):
Couple of minor comments. Looks good to me otherwise.

wentongwu commented on subsys/power/device.c:
```diff
@@ -57,7 +57,7 @@ int sys_pm_suspend_devices(void)
 						NULL, NULL);
 		if (device_retval[i]) {
-			LOG_ERR("%s suspend operation failed\n",
+			LOG_DBG("%s suspend operation failed\n",
```
@vanti Thanks for the review,  my commit message describe the reason ```
Current PM policy allow devices
make the decision if going to sleep/deep sleep state, so it's not error
message if failed suspend devices, change it to debug level.
``` 
and if using error message, it will confuse end users who are not familiar zephyr power management.

vanti replied:
In this case the message is too strong. When I see "failed" it sounds like an error message. How about "%s did not enter suspend state"?

wentongwu replied:
@vanti updated.

wentongwu commented on subsys/power/power.c:
```diff
@@ -116,6 +120,18 @@ enum power_states _sys_suspend(s32_t ticks)
+			/* low power peripherals. */
+			if (sys_pm_low_power_devices()) {
+				LOG_DBG("Device enter low power state failed!");
+				sys_pm_resume_devices();
```
@vanti sys_pm_suspend_devices will be in the environment of power gating most power domain, so devices can't work if system is in that state. But sys_pm_low_power_devices will be in state of clock gate or reduce clock, power is on, devices can work in that case, so even fail to low power devices, we still can have system enter sleep state as the same behavior before this change.

vanti replied:
I don't think it is strictly true that devices can work for any sleep state, at least per my understanding thus far. For instance, on the TI CC1352, there is a sleep state where it can enter a standby mode in which the CPU does not entirely lose its context (ie. it can resume from the point where it went to sleep), but some devices, such as UART/I2C/SPI do not work.  

That said, I expect this code would still work for CC1352, because its implementation of 'set_power_state' does not return failure. But I wonder if we are making an assumption that would cause an error to be masked on other platforms.

wentongwu replied:
what I thought is when CPU core only is in clock gate state, some devices can do hardware transfer (such as DMA can still do data transfer because before enter sleep state, someone gives DMA descriptor), when done the transfer, interrupt will come and wake-up CPU. If there is no that case, I will change the code. @vanti

wentongwu replied:
@vanti Maybe I should consider more, just be consistent with deep sleep currently, updated code now.

wentongwu committed:
power: active all devices if not all devices enter suspend state

Active all devices if not all devices enter suspend state when system
decide going to suspend state.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

vanti reviewed (approved):
Thanks, LGTM

nashif commented on subsys/power/device.c:
```diff
@@ -57,7 +57,7 @@ int sys_pm_suspend_devices(void)
 						NULL, NULL);
 		if (device_retval[i]) {
-			LOG_ERR("%s suspend operation failed\n",
+			LOG_DBG("%s did not enter suspend state\n",
```
remove that "\n"

wentongwu replied:
@nashif updated

wentongwu committed:
power: correct log level for power management

Current PM policy allow devices make the decision if going to
sleep/deep sleep state, so it's not error message if some
devices don't enter suspend state, change it to debug level.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
power: add DEVICE_PM_LOW_POWER_STATE for device power management

When system going to sleep state, make peripherals go to state
DEVICE_PM_LOW_POWER_STATE which needs less time than state
DEVICE_PM_SUSPEND_STATE to save more power.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
